docs/specs/biometric.md

```diff
@@ -0,0 +1,93 @@
+The Mobile Security Key is stored in the keychain on the device.
+Biometric authentication is required to retrieve the Mobile Security Key.
+
+The HSBC app offers a logout button in the app.
+If the user logs out, the current session is terminated, but biometric authentication is still available.
+It is suspected the HSBC app remembers the last login user.
+Next time the app is launched, the user can just use biometric authentication to login.
+
+It makes sense for a banking application to keep session lifetime short.
+
+Once a user has logged in the HSBC app, the app remembers the user forever.
+It is impossible to switch user unless the app is uninstalled and installed again.
+The uninstallation wipes the keychain.
+
+## Design Overview
+
+Storing the refresh token implies long session lifetime.
+Storing the password poses a hard restriction on the authenticator the user can use.
+In order to have short session lifetime, as well as support for passwordless login, OAuth or any other login means,
+we must store something unique.
+
+Borrowing the design of anonymous user, a keypair is stored in the keychain.
+Biometric authentication is required to retrieve the private key.
+The keypair is used to both identify and authenticate the user.
+
+## Security Concerns
+
+The common user experience of biometric authentication is by scanning fingerprint or face to
+log in an account which was signed in previously with non-biometric means.
+
+This means enabling biometric authentication binds biometric authentication to the application authentication.
+If the application is not a single-user application like the HSBC app,
+then we must ensure when User B logs in, User B can never access the keypair of User A.
+
+The SDK stores the following things:
+
+1. The keypair is stored in the device keychain per container, requiring biometric authentication to access.
+2. The last logged in user ID (LastUserID) is stored in the device keychain per container, NOT requiring biometric authentication to access.
+3. The user ID of last biometric setup (BiometricUserID) is stored in the device keychain per container, NOT requiring biometric authentication to access.
+
+Under normal condition, when biometric authentication is enabled, the user should log in with it.
+For some reasons like the user is wearing a mask or gloves, biometric authentication is impossible,
+the application should perform the typical web-based authentication flow to authenticate the user.
+It is possible that the user could log into a different account, which can lead to the situation that
+User B is logged in and the keypair of User A is still in the keychain.
+
+The HSBC app avoid this situation by making itself a single-user application.
+However, it is impossible for the SDK to enforce this.
+
+Instead, the SDK has the following behavior.
+If the BiometricUserID is not null, the SDK includes the BiometricUserID as `login_hint` in the authorization URL.
```

Comment:
I'm skeptical if it is really necessary, and may cause more confusion.

Imagine there could be multiple ways the UX of what you describe can be designed:

1. Biometrics login if enabled, if failed, switch back to page page where you can choose to try Biometrics login again, or forgot your account (in this way, no need the behavior you described)
2. Biometrics login if enabled, if failed, display the normal web login page. However, there are no way the developer can allow the user to disable the biometrics here again? Since it is already redirected to our login page; In this page, if user want to login to another account, what can he do? There are no obvious way to do it, a deadlock and UX problem.

Can we instead, just make our SDK's `authorize({ ... })` always clear the biometrics key if there is one?

Reply:
> Isn't that no matter what, if isBiometricEnabled() returns false, the developer knows biometric was not setup? How is it related if getLastUserID() return something or not?

It isBiometricEnabled() returns true, then getLastUserID() must return something.
If getLastUserID() returns nothing, then isBiometricEnabled() must return false.

> How should that be implemented? The developer should store the biometrics's user's Login ID or any meta data they want? is it possible?

This is only possible if we include all identities data in the user info endpoint. This is NOT yet done.

> Biometrics login if enabled, if failed, switch back to page page where you can choose to try Biometrics login again, or forgot your account (in this way, no need the behavior you described)

Failed for what reason? It is a failure due to covered face, then a retry on biometric authentication should work. Otherwise fallback to webview login limited to the last user.

> Biometrics login if enabled, if failed, display the normal web login page. However, there are no way the developer can allow the user to disable the biometrics here again? Since it is already redirected to our login page; In this page, if user want to login to another account, what can he do? There are no obvious way to do it, a deadlock and UX problem.

Xero shows "Sign in", which translates to web login limited to the last user. "Forget [email login ID]" translated to calling `disableBiometric()` followed by `authorize()`.

Reply:
> Xero shows "Sign in", which translates to web login limited to the last user. "Forget [email login ID]" translated to calling disableBiometric() followed by authorize().

Actually this is not true, the "Sign in" actually called `authorizeBiometric()` again in Xero's case.

So go back to our original discussion, what do you think if we just by default make `authorize({ ... })` always clear the biometrics key if there is one?